File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckSelectionDialog.java
/*
 Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 3 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.anki.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.ichi2.anki.R;
import com.ichi2.anki.UIUtils;
import com.ichi2.anki.analytics.AnalyticsDialogFragment;
import com.ichi2.libanki.Collection;
import com.ichi2.libanki.CollectionGetter;
import com.ichi2.libanki.Deck;
import com.ichi2.libanki.DeckManager;
import com.ichi2.libanki.backend.exception.DeckRenameException;
import com.ichi2.libanki.stats.Stats;
import com.ichi2.utils.DeckNameComparator;
import com.ichi2.utils.FunctionalInterfaces;
import com.ichi2.utils.FilterResultsUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import timber.log.Timber;

/**
 * The dialog which allow to select a deck. It is opened when the user click on a deck name in stats, browser or note editor.
 * It allows to filter decks by typing part of its name.
 */
public class DeckSelectionDialog extends AnalyticsDialogFragment {

    private MaterialDialog mDialog;


    /**
     * A dialog which handles selecting a deck
     */
    @NonNull
    public static DeckSelectionDialog newInstance(@NonNull String title, @Nullable String summaryMessage, @NonNull boolean keepRestoreDefaultButton, @NonNull List<SelectableDeck> decks) {
        DeckSelectionDialog f = new DeckSelectionDialog();
        Bundle args = new Bundle();
        args.putString("summaryMessage", summaryMessage);
        args.putString("title", title);
        args.putBoolean("keepRestoreDefaultButton", keepRestoreDefaultButton);
        args.putParcelableArrayList("deckNames", new ArrayList<>(decks));
        f.setArguments(args);
        return f;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(true);
    }


    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        View dialogView = LayoutInflater.from(getActivity())
                .inflate(R.layout.deck_picker_dialog, null, false);


        TextView summary = dialogView.findViewById(R.id.deck_picker_dialog_summary);

        Bundle arguments = requireArguments();

        if (getSummaryMessage(arguments) == null) {
            summary.setVisibility(View.GONE);
        } else {
            summary.setVisibility(View.VISIBLE);
            summary.setText(getSummaryMessage(arguments));
        }

        RecyclerView recyclerView = dialogView.findViewById(R.id.deck_picker_dialog_list);
        recyclerView.requestFocus();

        RecyclerView.LayoutManager deckLayoutManager = new LinearLayoutManager(requireActivity());
        recyclerView.setLayoutManager(deckLayoutManager);

        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(), DividerItemDecoration.VERTICAL);
        recyclerView.addItemDecoration(dividerItemDecoration);

        List<SelectableDeck> decks = getDeckNames(arguments);
        DecksArrayAdapter adapter = new DecksArrayAdapter(decks);
        recyclerView.setAdapter(adapter);

        adjustToolbar(dialogView, adapter);

        MaterialDialog.Builder builder = new MaterialDialog.Builder(requireActivity())
                .neutralText(R.string.dialog_cancel)
                .customView(dialogView, false);

        if (arguments.getBoolean("keepRestoreDefaultButton")) {
            builder = builder.negativeText(R.string.restore_default).onNegative((dialog, which) -> onDeckSelected(null));
        }

        mDialog = builder.build();
        return mDialog;
    }


    @Nullable
    private String getSummaryMessage(Bundle arguments) {
        return arguments.getString("summaryMessage");
    }


    @NonNull
    private ArrayList<SelectableDeck> getDeckNames(Bundle arguments) {
        return Objects.requireNonNull(arguments.getParcelableArrayList("deckNames"));
    }


    @NonNull
    private String getTitle() {
        return Objects.requireNonNull(requireArguments().getString("title"));
    }


    private void adjustToolbar(View dialogView, DecksArrayAdapter adapter) {
        Toolbar toolbar = dialogView.findViewById(R.id.deck_picker_dialog_toolbar);

        toolbar.setTitle(getTitle());

        toolbar.inflateMenu(R.menu.deck_picker_dialog_menu);

        MenuItem searchItem = toolbar.getMenu().findItem(R.id.deck_picker_dialog_action_filter);
        SearchView searchView = (SearchView) searchItem.getActionView();

        searchView.setQueryHint(getString(R.string.deck_picker_dialog_filter_decks));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                searchView.clearFocus();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.getFilter().filter(newText);
                return true;
            }
        });

        MenuItem addDecks = toolbar.getMenu().findItem(R.id.deck_picker_dialog_action_add_deck);
        addDecks.setOnMenuItemClickListener(menuItem -> {
            // creating new deck without any parent deck
            showDeckDialog();
            return true;
        });
    }

    private void showSubDeckDialog(String parentDeckPath) {
        try {
            // create subdeck
            Long parentId = getDecks().id(parentDeckPath);
            CreateDeckDialog createDeckDialog = new CreateDeckDialog(requireActivity(), R.string.create_subdeck, CreateDeckDialog.DeckDialogType.SUB_DECK, parentId);
            createDeckDialog.setOnNewDeckCreated((id) -> {
                // a sub deck was created
                selectDeckWithDeckName(getDecks().name(id));
            });
            createDeckDialog.showDialog();
        } catch (DeckRenameException ex) {
            Timber.w(ex);
        }
    }

    private void showDeckDialog() {
        CreateDeckDialog createDeckDialog =  new CreateDeckDialog(requireActivity(), R.string.new_deck, CreateDeckDialog.DeckDialogType.DECK, null);
        createDeckDialog.setOnNewDeckCreated((id) -> {
            // a deck was created
            selectDeckWithDeckName(getDecks().name(id));
        });
        createDeckDialog.showDialog();
    }

    @NonNull
    protected CollectionGetter requireCollectionGetter() {
        return (CollectionGetter) requireContext();
    }

    @NonNull
    protected DeckManager getDecks() {
        return requireCollectionGetter().getCol().getDecks();
    }


    /**
     * Create the deck if it does not exists.
     * If name is valid, send the deck with this name to listener and close the dialog.
     */
    private void selectDeckWithDeckName(@NonNull String deckName) {
        try {
            long id = getDecks().id(deckName);
            SelectableDeck dec = new SelectableDeck(id, deckName);
            selectDeckAndClose(dec);
        } catch (DeckRenameException ex) {
            UIUtils.showThemedToast(requireActivity(), ex.getLocalizedMessage(getResources()), false);
        }
    }


    /**
     * @param deck deck sent to the listener.
     */
    protected void onDeckSelected(@Nullable SelectableDeck deck) {
        getDeckSelectionListener().onDeckSelected(deck);
    }


    @NonNull
    private DeckSelectionListener getDeckSelectionListener() {
        Activity activity = requireActivity();
        if (activity instanceof DeckSelectionListener) {
            return (DeckSelectionListener) activity;
        }

        Fragment parentFragment = getParentFragment();
        if (parentFragment instanceof DeckSelectionListener) {
            return (DeckSelectionListener) parentFragment;
        }

        throw new IllegalStateException("Neither activity or parent fragment were a selection listener");
    }


    /**
     * Same action as pressing on the deck in the list. I.e. send the deck to listener and close the dialog.
     */
    protected void selectDeckAndClose(@NonNull SelectableDeck deck) {
        onDeckSelected(deck);
        mDialog.dismiss();
    }

    protected void displayErrorAndCancel() {
        mDialog.dismiss();
    }


    public class DecksArrayAdapter extends RecyclerView.Adapter<DecksArrayAdapter.ViewHolder> implements Filterable {
        public class ViewHolder extends RecyclerView.ViewHolder {
            private final TextView mDeckTextView;
            public ViewHolder(@NonNull TextView ctv) {
                super(ctv);
                mDeckTextView = ctv;
                mDeckTextView.setOnClickListener(view -> {
                    String deckName = ctv.getText().toString();
                    selectDeckByNameAndClose(deckName);
                });

                mDeckTextView.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View view) {
                        // creating sub deck with parent deck path
                        showSubDeckDialog(ctv.getText().toString());
                        return true;
                    }
                });
            }


            public void setDeck(@NonNull SelectableDeck deck) {
                mDeckTextView.setText(deck.getName());
            }
        }

        private final ArrayList<SelectableDeck> mAllDecksList = new ArrayList<>();
        private final ArrayList<SelectableDeck> mCurrentlyDisplayedDecks = new ArrayList<>();

        public DecksArrayAdapter(@NonNull List<SelectableDeck> deckNames) {
            mAllDecksList.addAll(deckNames);
            mCurrentlyDisplayedDecks.addAll(deckNames);
            Collections.sort(mCurrentlyDisplayedDecks);
        }

        protected void selectDeckByNameAndClose(@NonNull String deckName) {
            for (SelectableDeck d : mAllDecksList) {
                if (d.getName().equals(deckName)) {
                    selectDeckAndClose(d);
                    return;
                }
            }
            displayErrorAndCancel();
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.deck_picker_dialog_list_item, parent, false);

            return new ViewHolder(v.findViewById(R.id.deck_picker_dialog_list_item_value));
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            SelectableDeck deck = mCurrentlyDisplayedDecks.get(position);
            holder.setDeck(deck);
        }

        @Override
        public int getItemCount() {
            return mCurrentlyDisplayedDecks.size();
        }


        @NonNull
        @Override
        public Filter getFilter() {
            return new DecksFilter();
        }

        /* Custom Filter class - as seen in http://stackoverflow.com/a/29792313/1332026 */
        private class DecksFilter extends Filter {
            private final ArrayList<SelectableDeck> mFilteredDecks;
            protected DecksFilter() {
                super();
                mFilteredDecks = new ArrayList<>();
            }

            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                mFilteredDecks.clear();
                ArrayList<SelectableDeck> allDecks = DecksArrayAdapter.this.mAllDecksList;
                if (constraint.length() == 0) {
                    mFilteredDecks.addAll(allDecks);
                } else {
                    final String filterPattern = constraint.toString().toLowerCase(Locale.getDefault()).trim();
                    for (SelectableDeck deck : allDecks) {
                        if (deck.getName().toLowerCase(Locale.getDefault()).contains(filterPattern)) {
                            mFilteredDecks.add(deck);
                        }
                    }
                }

                return FilterResultsUtils.fromCollection(mFilteredDecks);
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                ArrayList<SelectableDeck> currentlyDisplayedDecks = DecksArrayAdapter.this.mCurrentlyDisplayedDecks;
                currentlyDisplayedDecks.clear();
                currentlyDisplayedDecks.addAll(mFilteredDecks);
                Collections.sort(currentlyDisplayedDecks);
                notifyDataSetChanged();
            }
        }
    }


    public static class SelectableDeck implements Comparable<SelectableDeck>, Parcelable {
        /**
         * Either a deck id or ALL_DECKS_ID
         */
        private final long mDeckId;
        /**
         * Name of the deck, or localization of "all decks"
         */
        private final String mName;


        /**
         * @param filter A method deciding which deck to add
         * @return the list of all SelectableDecks from the collection satisfying filter
         */
        @NonNull
        public static List<SelectableDeck> fromCollection(@NonNull Collection c, @NonNull FunctionalInterfaces.Filter<Deck> filter) {
            List<Deck> all = c.getDecks().all();
            List<SelectableDeck> ret = new ArrayList<>(all.size());
            for (Deck d : all) {
                if (!filter.shouldInclude(d)) {
                    continue;
                }
                ret.add( new SelectableDeck(d));
            }
            return ret;
        }

        @SuppressWarnings("unused")
        @NonNull
        public static List<SelectableDeck> fromCollection(@NonNull Collection c) {
            return fromCollection(c, FunctionalInterfaces.Filters.allowAll());
        }


        public SelectableDeck(long deckId, @NonNull String name) {
            this.mDeckId = deckId;
            this.mName = name;
        }


        protected SelectableDeck(@NonNull Deck d) {
            this(d.getLong("id"), d.getString("name"));
        }


        protected SelectableDeck(@NonNull Parcel in) {
            mDeckId = in.readLong();
            mName = in.readString();
        }


        public long getDeckId() {
            return mDeckId;
        }


        @NonNull
        public String getName() {
            return mName;
        }


        /** "All decks" comes first. Then usual deck name order. */
        @Override
        public int compareTo(@NonNull SelectableDeck o) {
            if (this.mDeckId == Stats.ALL_DECKS_ID){
                if (o.mDeckId == Stats.ALL_DECKS_ID) {
                    return 0;
                }
                return -1;
            }
            if (o.mDeckId == Stats.ALL_DECKS_ID) {
                return 1;
            }
            return DeckNameComparator.INSTANCE.compare(this.mName, o.mName);
        }


        @Override
        public int describeContents() {
            return 0;
        }


        @Override
        public void writeToParcel(@NonNull Parcel dest, int flags) {
            dest.writeLong(mDeckId);
            dest.writeString(mName);
        }



        @SuppressWarnings("unused")
        public static final Parcelable.Creator<SelectableDeck> CREATOR = new Parcelable.Creator<SelectableDeck>() {
            @Override
            public SelectableDeck createFromParcel(Parcel in) {
                return new SelectableDeck(in);
            }


            @Override
            public SelectableDeck[] newArray(int size) {
                return new SelectableDeck[size];
            }
        };
    }


    public interface DeckSelectionListener {
        void onDeckSelected(@Nullable SelectableDeck deck);
    }
}

File: AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DeckSelectionDialog.kt
/*
 Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 3 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.ichi2.anki.dialogs

import android.app.Activity
import android.app.Dialog
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageButton
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.os.BundleCompat
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import anki.decks.deckTreeNode
import com.ichi2.anki.ALL_DECKS_ID
import com.ichi2.anki.CollectionManager.withCol
import com.ichi2.anki.OnContextAndLongClickListener.Companion.setOnContextAndLongClickListener
import com.ichi2.anki.R
import com.ichi2.anki.analytics.AnalyticsDialogFragment
import com.ichi2.anki.common.annotations.NeedsTest
import com.ichi2.anki.common.utils.annotation.KotlinCleanup
import com.ichi2.anki.databinding.DeckPickerDialogListItemBinding
import com.ichi2.anki.databinding.DialogDeckPickerBinding
import com.ichi2.anki.dialogs.DeckSelectionDialog.DecksArrayAdapter.DecksFilter
import com.ichi2.anki.launchCatchingTask
import com.ichi2.anki.libanki.DeckId
import com.ichi2.anki.libanki.DeckNameId
import com.ichi2.anki.libanki.sched.DeckNode
import com.ichi2.anki.model.SelectableDeck
import com.ichi2.ui.AccessibleSearchView
import com.ichi2.utils.TypedFilter
import com.ichi2.utils.create
import com.ichi2.utils.customView
import com.ichi2.utils.negativeButton
import com.ichi2.utils.positiveButton
import timber.log.Timber
import java.util.Locale

/**
 * "Deck Search": A dialog allowing the user to select a deck from a list of decks.
 *
 * * Allows filtering of visible decks based on name (searching): [DecksFilter]
 * * Allows adding a new deck: [showDeckDialog]
 * * Allows adding a subdeck via long-pressing a deck: [showSubDeckDialog]
 *
 * It is opened when the user wants a deck in stats, browser or note editor.
 *
 * @see com.ichi2.anki.model.SelectableDeck The data that is displayed
 */
@NeedsTest("simulate 'don't keep activities'")
@NeedsTest("Test the ordering of the dialog")
@NeedsTest("test the ordering of decks in search page in the dialog")
@NeedsTest("test syncing the status of collapsing deck with teh deckPicker")
open class DeckSelectionDialog : AnalyticsDialogFragment() {
    private lateinit var binding: DialogDeckPickerBinding
    private var dialog: AlertDialog? = null
    private lateinit var expandImage: Drawable
    private lateinit var collapseImage: Drawable
    private lateinit var decksRoot: DeckNode

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCancelable = true

        val attrs =
            intArrayOf(
                R.attr.expandRef,
                R.attr.collapseRef,
            )
        val typedArray = requireContext().obtainStyledAttributes(attrs)
        expandImage = typedArray.getDrawable(0)!!
        expandImage.isAutoMirrored = true
        collapseImage = typedArray.getDrawable(1)!!
        collapseImage.isAutoMirrored = true
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        binding = DialogDeckPickerBinding.inflate(LayoutInflater.from(context))
        val summary = binding.summary
        val arguments = requireArguments()
        if (getSummaryMessage(arguments) == null) {
            summary.visibility = View.GONE
        } else {
            summary.visibility = View.VISIBLE
            summary.text = getSummaryMessage(arguments)
        }
        binding.list.requestFocus()
        val deckLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(requireActivity())
        binding.list.layoutManager = deckLayoutManager
        val dividerItemDecoration = DividerItemDecoration(binding.list.context, DividerItemDecoration.VERTICAL)
        binding.list.addItemDecoration(dividerItemDecoration)
        val decks: List<SelectableDeck> = getDeckNames(arguments)
        val adapter = DecksArrayAdapter(decks)
        binding.list.adapter = adapter
        adjustToolbar(binding.root, adapter)
        dialog =
            AlertDialog.Builder(requireActivity()).create {
                negativeButton(R.string.dialog_cancel)
                customView(view = binding.root)
                if (arguments.getBoolean(KEEP_RESTORE_DEFAULT_BUTTON)) {
                    positiveButton(R.string.restore_default) {
                        onDeckSelected(null)
                    }
                }
            }
        return dialog!!
    }

    override fun onResume() {
        super.onResume()
        dialog?.window?.clearFlags(
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_ALT_FOCUSABLE_IM,
        )
    }

    private fun getSummaryMessage(arguments: Bundle): String? = arguments.getString(SUMMARY_MESSAGE)

    private fun getDeckNames(arguments: Bundle): ArrayList<SelectableDeck> =
        BundleCompat.getParcelableArrayList(arguments, DECK_NAMES, SelectableDeck::class.java)!!

    private val title: String
        get() = requireArguments().getString(TITLE)!!

    private fun adjustToolbar(
        dialogView: View,
        adapter: DecksArrayAdapter,
    ) {
        val toolbar: Toolbar = binding.toolbar
        toolbar.title = title
        toolbar.inflateMenu(R.menu.deck_picker_dialog_menu)
        val searchItem = toolbar.menu.findItem(R.id.deck_picker_dialog_action_filter)
        val searchView = searchItem.actionView as AccessibleSearchView
        searchView.queryHint = getString(R.string.deck_picker_dialog_filter_decks)
        searchView.setOnQueryTextListener(
            object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    searchView.clearFocus()
                    return true
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    adapter.filter.filter(newText)
                    return true
                }
            },
        )
        val addDecks = toolbar.menu.findItem(R.id.deck_picker_dialog_action_add_deck)
        addDecks.setOnMenuItemClickListener {
            // creating new deck without any parent deck
            showDeckDialog()
            true
        }
    }

    /**
     * Displays a dialog to create a subdeck under the specified parent deck.
     *
     * @param parentDeck The parent deck under which the subdeck will be created
     */
    private fun showSubDeckDialog(parentDeck: SelectableDeck.Deck) {
        val createDeckDialog =
            CreateDeckDialog(requireActivity(), R.string.create_subdeck, CreateDeckDialog.DeckDialogType.SUB_DECK, parentDeck.deckId)
        createDeckDialog.onNewDeckCreated = { did: DeckId -> onNewDeckCreated(did) }
        createDeckDialog.showDialog()
    }

    private fun showDeckDialog() {
        val createDeckDialog = CreateDeckDialog(requireActivity(), R.string.new_deck, CreateDeckDialog.DeckDialogType.DECK, null)
        createDeckDialog.onNewDeckCreated = { did: DeckId -> onNewDeckCreated(did) }
        createDeckDialog.showDialog()
    }

    /** Updates the list and simulates a click on the newly created deck */
    private fun onNewDeckCreated(id: DeckId) {
        // a deck/subdeck was created
        launchCatchingTask {
            val name = withCol { decks.name(id) }
            val deck = SelectableDeck.Deck(id, name)
            deckCreationListener?.onDeckCreated(DeckNameId(name, id))
            selectDeckAndClose(deck)
        }
    }

    /**
     * @param deck deck sent to the listener.
     */
    private fun onDeckSelected(deck: SelectableDeck?) {
        deckSelectionListener!!.onDeckSelected(deck)
    }

    @KotlinCleanup("Use a factory here")
    var deckSelectionListener: DeckSelectionListener? = null
        get() {
            if (field != null) {
                return field
            }
            val parentFragment = parentFragment
            if (parentFragment is DeckSelectionListener) {
                return parentFragment
            }
            val activity: Activity = requireActivity()
            if (activity is DeckSelectionListener) {
                return activity
            }
            // try to find inside the activity an active fragment that is a DeckSelectionListener
            val foundAvailableFragments =
                parentFragmentManager.fragments.filter {
                    it.isResumed && it is DeckSelectionListener
                }
            if (foundAvailableFragments.isNotEmpty()) {
                // if we found at least one resumed candidate fragment use it
                return foundAvailableFragments[0] as DeckSelectionListener
            }
            throw IllegalStateException("Neither activity or any fragment in the activity were a selection listener")
        }

    var deckCreationListener: DeckCreationListener? = null

    /**
     * Same action as pressing on the deck in the list. I.e. send the deck to listener and close the dialog.
     */
    protected fun selectDeckAndClose(deck: SelectableDeck) {
        Timber.d("selected deck '%s'", deck)
        onDeckSelected(deck)
        dialog!!.dismiss()
    }

    protected fun displayErrorAndCancel() {
        dialog!!.dismiss()
    }

    open inner class DecksArrayAdapter(
        decks: List<SelectableDeck>,
    ) : RecyclerView.Adapter<DecksArrayAdapter.ViewHolder>(),
        Filterable {
        inner class ViewHolder(
            private val binding: DeckPickerDialogListItemBinding,
        ) : RecyclerView.ViewHolder(binding.root) {
            private var currentDeck: SelectableDeck? = null

            val expander: ImageButton = binding.expander
            val indentView: ImageButton = binding.indent

            fun setDeck(deck: SelectableDeck) {
                binding.deckTextView.text = deck.getDisplayName(requireContext())
                currentDeck = deck
            }

            init {
                binding.root.setOnClickListener {
                    currentDeck?.let { selectDeckAndClose(it) }
                }
                expander.setOnClickListener {
                    currentDeck?.let { toggleExpansion(it) }
                }
                binding.root.setOnContextAndLongClickListener {
                    // creating sub deck with parent deck path
                    currentDeck?.let { deck ->
                        if (deck is SelectableDeck.Deck) {
                            showSubDeckDialog(deck)
                        }
                    }

                    true
                }
            }

            private fun toggleExpansion(deck: SelectableDeck) {
                val deckId =
                    when (deck) {
                        is SelectableDeck.AllDecks -> return
                        is SelectableDeck.Deck -> deck.deckId
                    }
                decksRoot.find(deckId)?.apply {
                    collapsed = !collapsed
                    Timber.d("The deck with ID $id is currently expanded: ${!collapsed}.")
                    updateCurrentlyDisplayedDecks()
                }
            }
        }

        private fun updateCurrentlyDisplayedDecks() {
            currentlyDisplayedDecks.clear()
            currentlyDisplayedDecks.addAll(allDecksList.filter(::isViewable))
            notifyDataSetChanged()
        }

        private val allDecksList = ArrayList<DeckNode>()
        private val currentlyDisplayedDecks = ArrayList<DeckNode>()

        override fun onCreateViewHolder(
            parent: ViewGroup,
            viewType: Int,
        ): ViewHolder {
            val layoutInflater = LayoutInflater.from(context)
            val binding = DeckPickerDialogListItemBinding.inflate(layoutInflater, parent, false)
            return ViewHolder(binding)
        }

        override fun onBindViewHolder(
            holder: ViewHolder,
            position: Int,
        ) {
            val deck = currentlyDisplayedDecks[position]
            val isDeckViewable = isViewable(deck)
            holder.itemView.isVisible = isDeckViewable
            if (isDeckViewable) {
                val model = if (deck.did == ALL_DECKS_ID) SelectableDeck.AllDecks else SelectableDeck.Deck(deck.did, deck.fullDeckName)
                holder.setDeck(model)
            }
            setDeckExpander(holder.expander, holder.indentView, deck)
        }

        /**
         * Sets the expander and indent views based on the properties of the provided DeckNode.
         *
         * @param expander The ImageButton used for expanding/collapsing the deck node.
         * @param indent The ImageButton used for indenting the deck node.
         * @param node The DeckNode representing the deck.
         */
        private fun setDeckExpander(
            expander: ImageButton,
            indent: ImageButton,
            node: DeckNode,
        ) {
            if (hasSubDecks(node)) {
                expander.apply {
                    importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_YES
                    setImageDrawable(if (node.collapsed) expandImage else collapseImage)
                    contentDescription = context.getString(if (node.collapsed) R.string.expand else R.string.collapse)
                    visibility = View.VISIBLE
                }
            } else {
                expander.apply {
                    visibility = View.INVISIBLE
                    importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO
                }
            }
            indent.minimumWidth = node.depth * expander.resources.getDimensionPixelSize(R.dimen.keyline_1)
        }

        private fun hasSubDecks(node: DeckNode): Boolean = node.children.isNotEmpty()

        private fun isViewable(deck: DeckNode): Boolean {
            val parentNodeRef = deck.parent ?: return true
            // The parent belongs to the tree retained by [allDecksList], so should still exist.
            val parentNode = parentNodeRef.get()!!
            return !parentNode.collapsed && isViewable(parentNode)
        }

        override fun getItemCount(): Int = currentlyDisplayedDecks.size

        override fun getFilter(): Filter = DecksFilter()

        private inner class DecksFilter : TypedFilter<DeckNode>(allDecksList) {
            override fun filterResults(
                constraint: CharSequence,
                items: List<DeckNode>,
            ): List<DeckNode> {
                val filterPattern = constraint.toString().lowercase(Locale.getDefault()).trim()
                return items.filter {
                    it.fullDeckName.lowercase(Locale.getDefault()).contains(filterPattern)
                }
            }

            override fun publishResults(
                constraint: CharSequence?,
                results: List<DeckNode>,
            ) {
                results.forEach { it.collapsed = false }
                currentlyDisplayedDecks.apply {
                    clear()
                    addAll(results)
                }
                notifyDataSetChanged()
            }
        }

        init {
            launchCatchingTask {
                decksRoot = withCol { Pair(sched.deckDueTree(), isEmpty) }.first
                val allDecksSet =
                    decks
                        .mapNotNull { it as? SelectableDeck.Deck }
                        .mapNotNull { decksRoot.find(it.deckId) }
                        .toSet()
                if (decks.any { it is SelectableDeck.AllDecks }) {
                    val newDeckNode =
                        deckTreeNode {
                            deckId = ALL_DECKS_ID
                            name = "all"
                        }
                    allDecksList.add(DeckNode(newDeckNode, getString(R.string.card_browser_all_decks), null))
                }

                allDecksList.addAll(allDecksSet)
                updateCurrentlyDisplayedDecks()
            }
        }
    }

    // TODO: allow filtering to SelectableDeck.Deck, excluding 'AllDecks'

    fun interface DeckSelectionListener {
        fun onDeckSelected(deck: SelectableDeck?)
    }

    fun interface DeckCreationListener {
        fun onDeckCreated(deck: DeckNameId)
    }

    companion object {
        private const val SUMMARY_MESSAGE = "summaryMessage"
        private const val TITLE = "title"
        private const val KEEP_RESTORE_DEFAULT_BUTTON = "keepRestoreDefaultButton"
        private const val DECK_NAMES = "deckNames"

        /**
         * A dialog which handles selecting a deck
         */
        fun newInstance(
            title: String,
            summaryMessage: String?,
            keepRestoreDefaultButton: Boolean,
            decks: List<SelectableDeck>,
        ): DeckSelectionDialog {
            val f = DeckSelectionDialog()
            val args = Bundle()
            args.putString(SUMMARY_MESSAGE, summaryMessage)
            args.putString(TITLE, title)
            args.putBoolean(KEEP_RESTORE_DEFAULT_BUTTON, keepRestoreDefaultButton)
            args.putParcelableArrayList(DECK_NAMES, ArrayList(decks))
            f.arguments = args
            return f
        }
    }
}
